flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/ReconciliationUtils.java

```diff
@@ -229,24 +227,32 @@ private static boolean upgradeStarted(
                 || currentReconState == ReconciliationState.UPGRADING;
     }
 
-    public static <T> T deserializedSpecWithVersion(
+    public static <T> Tuple2<T, ObjectNode> deserializeSpecWithMeta(
             @Nullable String specString, Class<T> specClass) {
         if (specString == null) {
             return null;
         }
 
         try {
             ObjectNode objectNode = (ObjectNode) objectMapper.readTree(specString);
+            ObjectNode internalMeta = (ObjectNode) objectNode.remove(INTERNAL_METADATA_JSON_KEY);
+
+            // backward compatibility
             objectNode.remove("apiVersion");
-            return objectMapper.treeToValue(objectNode, specClass);
+
+            return Tuple2.of(objectMapper.treeToValue(objectNode, specClass), internalMeta);
         } catch (JsonProcessingException e) {
             throw new RuntimeException("Could not deserialize spec, this indicates a bug...", e);
         }
     }
 
-    public static String writeSpecWithCurrentVersion(Object spec) {
+    public static String writeSpecWithMeta(
```

Comment:
How about introducing a wrapper object here instead of modifying the objectnode?

Reply:
makes sense